MAINT: interpolate: speed up the RBFInterpolator evaluation with high dimensional interpolants

segasai opened the pull request:
#### Reference issue
Closes: #15150 

#### What does this implement/fix?
This moves out the matrix multiplication out of pythran code and instead uses np.dot which is noticeably faster
due to blas. Because this code structure may potentially lead to more memory use, the evaluation is being split into chunks if needed.

The speed up is small for low dimensional interpolants (<10%), but reaches almost factor of 40  for interpolants with 4000 dimensions.

#### Additional information
The performance comparison. 
Here is the comparison of the time required to evaluate the model vs number of dimensions of the interpolant (4,40,400,4000).
The first number is the performance of hand-coded Rbf() evaluation. The second one is the performance of RBFInterpolator. 
Performance of patched code:
```
Ndim TimeRbf TimeRBFInterpolator
4 0.071  0.088
40 0.076  0.086
400 0.128  0.1430
4000 0.592  0.604
```
Performance of original code: 
```
Ndim TimeRbf TimeRBFInterpolator
4 0.071 0.098
40 0.076 0.320
400 0.128 2.605
4000 0.576 27.779
```
This is the code doing the testing (executed using 
`
env OMP_NUM_THREADS=1  python test.py 
`)

The code that does testing:

```python
import numpy as np
import scipy.interpolate
import time

rstate = np.random.default_rng(44)


def doit(nx, neval=1000, ndim=2, npt=6000):
    # initial point distribution
    xs = rstate.normal(size=(ndim, npt))
    xgrid = np.arange(nx)
    data = (xgrid[None, :] * xs[0][:, None]**2 + xs[-1][:, None]**2)

    # number of points
    xnew = rstate.normal(size=(ndim, neval))

    # use the old interface on 1d data to get coefficients to apply for the
    # to the 4000d data vectors
    RR = scipy.interpolate.Rbf(*([_ for _ in xs] + [data[:, 0]]))
    coeffs = scipy.linalg.solve(RR.A, data)
    t1 = time.time()
    r = RR._call_norm(xnew, RR.xi)
    pred0 = np.dot(RR._function(r), coeffs)
    t20 = time.time()

    # use the new interface
    RR = scipy.interpolate.RBFInterpolator(xs.T, data)
    t21 = time.time()
    pred1 = RR(xnew.T)
    t3 = time.time()

    return (t20 - t1, t3 - t21)


print('Ndim TimeRbf TimeRBFInterpolator')
for i in [4, 40, 400, 4000]:
    res = doit(i)
    print('%d %.3f %.3f' % (i, res[0], res[1]))
```

tupui commented on scipy/interpolate/_rbfinterp.py:
```diff
@@ -10,27 +10,17 @@
+}
 
 # The shape parameter does not need to be specified when using these RBFs.
 _SCALE_INVARIANT = {"linear", "thin_plate_spline", "cubic", "quintic"}
```
You probably used autoformating on the whole file. Please don't as it not good for the history and does not facilitate reviews. If there are clear pep8 violations around the changes it's fine, but not more like this one for instance where it's very subjective.

segasai replied:
Yes, that was my mistake, I've forgot to turn yapf off, and then it too late with too many changes in the file.

tupui replied:
Please revert these.

tupui replied:
Thank you 🙏

tupui reviewed (changes requested):
Thanks @segasai. Just a generic comment. @treverhines can you have a look? 😃

tupui commented on scipy/interpolate/_rbfinterp.py:
```diff
@@ -387,6 +387,38 @@ def __init__(self, y, d,
         self.epsilon = epsilon
         self.powers = powers
 
+    def __chunk_evaluator(self,
+                          x,
+                          y,
+                          shift,
+                          scale,
+                          coeffs,
+                          memory_budget=None):
```
Any reason for mangling the name? Just marking it private should be enough.

Also please describe the memory option.

For the indentation. The following is preferred as it minimizes diffs. Same applies bellow in functions/methods usages.
```suggestion
    def _chunk_evaluator(
        self,
        x,
        y,
        shift,
        scale,
        coeffs,
        memory_budget=None
    ):
```

segasai replied:
done

treverhines commented on scipy/interpolate/_rbfinterp.py:
```diff
@@ -387,6 +387,64 @@ def __init__(self, y, d,
+        else:
+            nnei = self.neighbors
+        # in each chunk we consume the same space we already occupy
+        chunksize = memory_budget // ((self.powers.shape[0] + nnei)) + 1
```
Setting `chunksize` so that each chunk is about the same size as the class is not optimal when you are trying to interpolate a small number of observations at many evaluation points. My preference would be to have chunksize hard-coded (`chunksize=1024` seems to work well for me).

segasai replied:
I have set up the minimum memory budget to be 1e6, so that should prevent I think too small chunk sizes.

treverhines commented on scipy/interpolate/_rbfinterp.py:
```diff
@@ -387,6 +387,64 @@ def __init__(self, y, d,
+            for i in range(0, nx, chunksize):
+                # I had to use copy() here because pythran doesnt seem to
+                # accept views
+                vec = _evaluate(x[i:i + chunksize, :].copy(), y, self.kernel,
```
the copies are not needed. Pythran just throws an error when the array is not contiguous. `x` is c contiguous, so you can pass its slices to `_evaluate` without issue.  `coeffs` is fortran-contiguous, so pythran does throw an error when you slice it. However, `_evaluate` no longer users `coeffs`, so there is no longer any need to pass it to `_evaluate`.

segasai replied:
fixed in new version

treverhines commented on scipy/interpolate/_rbfinterp_pythran.py:
```diff
@@ -173,7 +173,7 @@ def _build_system(y, d, smoothing, kernel, epsilon, powers):
 #                          float[:],
 #                          float[:, :])
 def _evaluate(x, y, kernel, epsilon, powers, shift, scale, coeffs):
```
This function should be renamed and the doc strings should be updated since it no longer evaluates the interpolant. Perhaps rename it to `_build_evaluation_matrix`? Also, `coeffs` should be removed since it is no longer used.

segasai replied:
fixed in new version

treverhines commented on scipy/interpolate/_rbfinterp.py:
```diff
@@ -407,17 +465,19 @@ def __call__(self, x):
-                self._scale, self._coeffs
-                )
-
+            out = self._chunk_evaluator(
```
It is a bit awkward and confusing that `_chunk_evaluator` is a method, but you still have to pass attributes in as arguments. My preference would be to have `_chunk_evaluator` as a function just like what `_evaluate` used to be.

segasai replied:
The are several common class attributes that are used inside chunk_evaluator, so I think the current interface make sense to me.

treverhines commented on scipy/interpolate/_rbfinterp.py:
```diff
@@ -387,6 +387,64 @@ def __init__(self, y, d,
+            Domain shift used to create the polynomial matrix.
+        scale : (N,) float ndarray
+            Domain scaling used to create the polynomial matrix.
+        coeffs: (Q, P+R) float ndarray
```
The shape of coeffs is (P + R, S), where R is the number of monomials and S is the number of data dimensions.

segasai replied:
fixed

treverhines reviewed (commented):
I tested this out a bit and made a few comments.  The speedup for high-dimensional data is huge.  However, I am seeing some regressions in the benchmarks for this class (the benchmarks are only interpolating scalar data). It is about 50% slower when `n_samples` is 10 and 100 (`n_samples` is both the number of observation points and interpolation points).  I believe this can be fixed by using a larger chunk size.

segasai commented:
Thanks for the testing and review @treverhines . I'll try to address most of the issues. Regarding the chunk sizes, I think I'll change it to make memory_budget to be a minimum of 100Mb or something like that, which should make the chunk_size much bigger for interpolation of small datasets on large number of datapoints.

segasai commented:
Hi, 
I just wanted to ping on this PR. 
I think as far as I know I addressed all the comments/suggested changes.  
It would be great if @treverhines  or @tupui could have a look again to see if this is committable. 

Thanks in advance!

ev-br commented on scipy/interpolate/_rbfinterp.py:
```diff
@@ -387,6 +389,73 @@ def __init__(self, y, d,
+            array of points on which to evaluate
+        y: (P, N) float ndarray
+            array of points on which we know function values
+        shift: (N, ) array
```
nit: sometimes it's `array`, sometimes it's `ndarray`. Did you mean `array_like` (i.e., can be a list and you wrap it in an `np.asarray`)? OTOH, it's an internal function, so this is not high-prio

segasai replied:
I've changed the comment to ndarray (without trying the np.asarray conversions)

ev-br commented on scipy/interpolate/_rbfinterp.py:
```diff
@@ -387,6 +389,73 @@ def __init__(self, y, d,
+        else:
+            nnei = self.neighbors
+        # in each chunk we consume the same space we already occupy
+        chunksize = memory_budget // ((self.powers.shape[0] + nnei)) + 1
```
The default `memory_budget` is None?

segasai replied:
This was the decision to let the caller to always make the decision on the memory_budget. I.e. the default memory budget is set in the top level functions based on the total size of arrays. I'm sure it can be done differently, but it made sense to me.

ev-br replied:
It's fine in principle, the issue is that None should likely be replaced by a numerical value

segasai replied:
I can put the default value to 1e6 if you think it's better. The disadvantage from my poiint of view is that the 1e6 value will be in now two locations in the code.

segasai replied:
okay, I've put the default to 1e6

ev-br reviewed (commented):
Apologies it took a bit long to review. Looks good module a couple of minor nits.
Would be nice to test chunking a bit, esp where the data length and the chunk size are not commensurate so that the last chunk is shorter then the rest. Would you be able to add something of this sort @segasai ?

segasai commented on scipy/interpolate/_rbfinterp.py:
```diff
@@ -387,6 +389,73 @@ def __init__(self, y, d,
+            memory_budget=None
+    ):
+        """
+        Evaluate the interpolation why controlling memory consumption.
```
fixed

segasai commented:
Thanks fort taking a look @ev-br. I've added a test wth a big input array that should exercise the chunking where the last chunk should not be full.

ev-br commented:
The CI failures do not seem related, review comments are all addressed, including earlier comments by Pamphille, all in all it seems to be a great improvement --- merged. Thanks @segasai , all.

ev-br commented:
Would you please add a note for for the 1.9.0 release notes @segasai : https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.9.0

ev-br commented on scipy/interpolate/_rbfinterp.py:
```diff
@@ -396,7 +396,7 @@ def _chunk_evaluator(
             scale,
             coeffs,
-            memory_budget=None
+            memory_budget=1000000
```
I'd typically go for `if memory_budget is None: memory_budget=int(1e6)` but this is really minor and is not worth the churn unless you want to fix in a sequel (I'm fine either way)

segasai commented:
Thank you @ev-br  for merging/review! (and others who reviewed earlier)
I've added a line to the release notes about the change.
